File: 665.non-decreasing-array.js
/*
 * @lc app=leetcode id=665 lang=javascript
 *
 * [665] Non-decreasing Array
 */

// @lc code=start
/**
 * @param {number[]} nums
 * @return {boolean}
 */
var checkPossibility = function(nums) {
    if (nums.length == 0) {
        return true
    }

    let flipped = false;
    let prev = null;

    // note: don't use foreach when needing to break xDD
    // nums.forEach(function(elem, ind) { // to break a foreach loop need to throw an exception
    
    // nums.some(elem => {

    for (let elem of nums.keys()) {
        console.log(prev, nums[elem], flipped)
        if (prev == null) {
            prev = nums[elem]
        }
        else {        
            if (prev > nums[elem - 1] && flipped == true) {
                console.log("trrruuuuuuuuuuue")
                return false;
            }
            else if (prev > nums[elem] && flipped == false) {
                flipped = true
                nums.splice(elem - 1, 1)
                console.log(nums)
                // elem = prev + 1
            }
            // prev = elem
        }
    // });
    }
    
    const sorted = [...nums].sort()
    console.log(sorted, nums)
    for (let i = 0; i < nums.length; i++) {
        if (sorted[i] != nums[i]) {
            return false
        }
    }
    return true

};
// @lc code=end

File: 665.non-decreasing-array.py
#
# @lc app=leetcode id=665 lang=python3
#
# [665] Non-decreasing Array
#

# @lc code=start
class Solution:
    def checkPossibility(self, nums: List[int]) -> bool:
        if len(nums) == 0 or len(nums) == 1:
            return True
        
        prev = None
        flipped = None
        removeFirst = False
        removePrev = False
        for i in range(len(nums)):
            if prev == None:
                prev = nums[i]
            elif prev > nums[i] and flipped == None:
                # print("1")
                flipped = True

                removeFirst = nums[:i] + nums[i+1:]
                # print(removeFirst, sorted(removeFirst))
                if sorted(removeFirst) == removeFirst:
                    # print("isSorted removeFirst")
                    removeFirst = True
                    break

                removePrev = nums[:i-1] + nums[i:]
                # print(removePrev, sorted(removePrev))
                if sorted(removePrev) == removePrev:
                    # print("isSorted removePrev")
                    removePrev = True
                    break

                # print(removeFirst, removePrev)
                if removeFirst == True or removePrev == True:
                    return True
                return False
            prev = nums[i]

        # at least one must be sorted
        # iterate through rest of elements
        # print(removeFirst, sorted(removeFirst))
        if flipped is None:
            return True
        # print("end", removeFirst, removePrev)
        if removeFirst or removePrev:
            return True
        return False


# @lc code=end

